xapian-letor/api/featurelist_internal.cc

```diff
@@ -0,0 +1,186 @@
+    for (Xapian::TermIterator qt = featurelist_query.get_terms_begin(); qt != featurelist_query.get_terms_end(); ++qt) {
+	docterms.skip_to(*qt);
+	if (docterms != featurelist_doc.termlist_end() && *qt == *docterms) {
+	    tf[*qt] = docterms.get_wdf();
+	} else {
+	    tf[*qt] = 0;
+	}
+    }
+    termfreq = tf;
+}
+
+void
+FeatureList::Internal::compute_inverse_doc_freq()
+{
+    std::map<std::string, double> idf;
+
+    for (Xapian::TermIterator qt = featurelist_query.get_terms_begin(); qt != featurelist_query.get_terms_end(); ++qt) {
+	if (featurelist_db.term_exists(*qt)) {
+	    long int totaldocs = featurelist_db.get_doccount();
+	    long int df = featurelist_db.get_termfreq(*qt);
+	    idf[*qt] = log10(totaldocs / (1 + df));
+	} else {
+	    idf[*qt] = 0;
+	}
+    }
+    inverse_doc_freq = idf;
+}
+
+void
+FeatureList::Internal::compute_doc_length()
+{
+    std::map<std::string, long int> len;
+
+    long int temp_count = 0;
+    Xapian::TermIterator dt = featurelist_doc.termlist_begin();
+    // reach the iterator to the start of the title terms i.e. prefix "S"
+    dt.skip_to("S");
+    for ( ; dt != featurelist_doc.termlist_end(); ++dt) {
+	if ((*dt)[0] != 'S') {
+	    // We've reached the end of the S-prefixed terms.
+	    break;
+	}
+	temp_count += dt.get_wdf();
+    }
+    len["title"] = temp_count;
+    len["whole"] = featurelist_db.get_doclength(featurelist_doc.get_docid());
+    len["body"] = len["whole"] - len["title"];
+    doc_length = len;
+}
+
+void
+FeatureList::Internal::compute_collection_length()
+{
+    std::map<std::string, long int> len;
+
+    if (!featurelist_db.get_metadata("collection_len_title").empty() &&
+	!featurelist_db.get_metadata("collection_len_body").empty() &&
+	!featurelist_db.get_metadata("collection_len_whole").empty()) {
+	len["title"] = atol(featurelist_db.get_metadata("collection_len_title").c_str());
+	len["body"] = atol(featurelist_db.get_metadata("collection_len_body").c_str());
+	len["whole"] = atol(featurelist_db.get_metadata("collection_len_whole").c_str());
+    } else {
+	long int temp_count = 0;
+	Xapian::TermIterator dt = featurelist_db.allterms_begin("S");
+	for ( ; dt != featurelist_db.allterms_end("S"); ++dt) {
+	    //  because we don't want the unique terms so we want their
+	    // original frequencies and i.e. the total size of the title collection.
+	    temp_count += featurelist_db.get_collection_freq(*dt);
+	}
+	len["title"] = temp_count;
+	len["whole"] = featurelist_db.get_avlength() * featurelist_db.get_doccount();
+	len["body"] = len["whole"] - len["title"];
+    }
+    collection_length = len;
+}
+
+void
+FeatureList::Internal::compute_collection_termfreq()
+{
+    std::map<std::string, long int> tf;
+
+    for (Xapian::TermIterator qt = featurelist_query.get_terms_begin(); qt != featurelist_query.get_terms_end(); ++qt) {
+	if (featurelist_db.term_exists(*qt))
+	    tf[*qt] = featurelist_db.get_collection_freq(*qt);
+	else
+	    tf[*qt] = 0;
```

Comment:
This `term_exists()` check is unnecessary and inefficient - `get_collection_freq()` returns zero for terms not in the DB.